services/monitor/src/PendingContract.ts

```diff
@@ -0,0 +1,165 @@
+import logger from "./logger";
+import { FileHash } from "./util";
+import { Metadata, MetadataSourceMap } from "@ethereum-sourcify/lib-sourcify";
+import { id as keccak256str } from "ethers";
+import { KnownDecentralizedStorageFetchers } from "./types";
+import assert from "assert";
+import dotenv from "dotenv";
+dotenv.config();
+
+export default class PendingContract {
+  public metadataHash: FileHash;
+  public address: string;
+  public chainId: number;
+  public metadata: Metadata | undefined;
+  public pendingSources: MetadataSourceMap = {};
+  public fetchedSources: MetadataSourceMap = {};
+  private decentralizedStorageFetchers: KnownDecentralizedStorageFetchers;
+
+  constructor(
+    metadataHash: FileHash,
+    address: string,
+    chainId: number,
+    decentralizedStorageFetchers: KnownDecentralizedStorageFetchers
+  ) {
+    this.metadataHash = metadataHash;
+    this.address = address;
+    this.chainId = chainId;
+    this.decentralizedStorageFetchers = decentralizedStorageFetchers;
+  }
+
+  /**
+   * Assembles this contract by first fetching its metadata and then fetching all the sources listed in the metadata.
+   */
+  assemble = async () => {
+    const metadataFetcher =
+      this.decentralizedStorageFetchers[this.metadataHash.origin];
+    if (!metadataFetcher) {
+      throw new Error(
+        `No metadata fetcher found for origin ${this.metadataHash.origin}`
+      );
+    }
+    const metadataStr = await metadataFetcher
+      .fetch(this.metadataHash)
+      .catch((err) => {
+        throw new Error(
+          `Can't fetch metadata for ${this.address} ${this.metadataHash.hash}: ${err}`
+        );
+      });
+    logger.info(
+      `Fetched metadata for ${this.address} on chain ${this.chainId} from ${this.metadataHash.origin}`
+    );
+    // TODO: check if metadata hash matches this.metadataHash.hash
```

Comment:
Why do we need this check? Because a potential source (ipfs, swarm) could not return the right file requested? If that's the case, I think we can skip the check and trust the source.

Reply:
Oh I see, for sources we can use the keccak but for the metadata file we must encode the metadata to ipfs or bzzr address right?

Reply:
Yes. let me add it